Add ratelimiting feature to KafkaConsumer

glaksh100 opened the pull request:
## What is the purpose of the change
This PR adds a rate-limiting feature to the `KafkaConsumerThread` by using Guava's [RateLimiter](https://google.github.io/guava/releases/19.0/api/docs/index.html?com/google/common/util/concurrent/RateLimiter.html). 

## Brief change log
  - The change is behind a feature flag which can be passed through the properties for the Kafka consumer — `kafka.consumer.use.ratelimit`.
  - The `bytesPerSecondMax` is set using a Kafka consumer property — `kafka.consumer.bytespersecondmax`.
  - A `rateLimiter` is created if the feature flag is set to `true`.
  - Bytes read / second is calculated as `size of records in bytes (key + value) / time taken to execute the run loop`.
  - The rate is then adjusted as `pollingFrequency * bytesPerSecondMax / bytesReadPerSecond` .

## Verifying this change
  - Added a test in `KafkaConsumerThreadTest` to verify the change.

## Does this pull request potentially affect one of the following parts:
  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation
  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / **JavaDocs** / not documented)

lrao100 committed:
Add ratelimiting feature to KafkaConsumer

Update property name

glaksh100 commented:
Sharing on the lyft fork initially to get feedback, particularly around correctness of approach. I will open a FLINK JIRA and create an upstream PR soon.

tweise commented:
Yeah, we should have an upstream story before merging this. But you can proceed with testing at scale by building the application jar locally anyways.

I would probably look to add an override capability first before trying to put a specific implementation into Flink.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -65,6 +68,18 @@
+	private static final String USE_RATELIMITING = "kafka.consumer.ratelimit";
+
+	/** Property that sets the max bytes per second per consumer. **/
+	private static final String BYTES_PER_SECOND_MAX = "kafka.consumer.bytespersecondmax";
```
Let's unify the config naming a bit.  How about `kafka.consumer.ratelimit.enabled` and `kafka.consumer.ratelimit.maxbytespersecond`.

glaksh100 replied:
Done.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -254,7 +285,12 @@ public void run() {
+						records = getRecordsFromKafka();
+						runLoopEndTimeNanos = System.nanoTime();
+						// If useRateLimiting is set to true, rate is updated
+						updateRate(records, runLoopStartTimeNanos, runLoopEndTimeNanos);
```
Why do we have to update the rate each time through the loop?

glaksh100 replied:
Updated.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -482,6 +518,95 @@ void reassignPartitions(List<KafkaTopicPartitionState<TopicPartition>> newPartit
+		if (useRateLimiting) {
+			long recordBatchSizeBytes = getRecordBatchSize(records);
+			long runLoopTimeNanos = runLoopEndTimeNanos - runLoopStartTimeNanos;
+			double adjustedRate = adjustPollingFrequency(runLoopTimeNanos,
```
I suspect we're making this too complicated by not letting Guava's RateLimiter do more of the heavy lifting for us.  I think the correct way to approach this is have the cost unit of the RateLimiter be bytes.  Then you call `rateLimiter.acquire(numBytesJustRead)` or something like that.  At the point when we're beginning to read too fast the `acquire()` call will block for a bit and slow things down.

Anyway, that's my impression.  In this implementation you're calculating a desired rate and doing the equivalent of acquire(1).  I think this could be much simpler.  You shouldn't have to measure the loop frequency or any of that I believe if you use the RateLimiter properly.

glaksh100 replied:
That makes so much sense. Thanks for recommending a simplification :)

lrao100 committed:
Make rateLimiter logic simpler and unify config properties

glaksh100 commented:
Post discussion with @jgrier , made this a lot simpler and removed unnecessary run loop time calculations and updated the unit test.

The logic is similar to what @jgrier suggested in this comment: https://github.com/lyft/flink/pull/21#discussion_r251636503

glaksh100 commented:
For the upstream story, I think the following steps should work:
- Remove all but `getRecordsFromKafka()` in `KafkaConsumerThread`
- Make `getRecordsFromKafka()` protected
- Move the `KafkaConsumerThread` [instantiation](https://github.com/lyft/flink/blob/master/flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/Kafka09Fetcher.java#L105) to a protected method in  Kafka09Fetcher

This along with a couple of `getter` methods will suffice I think.. unless we think we can merge this as is.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -482,6 +511,67 @@ void reassignPartitions(List<KafkaTopicPartitionState<TopicPartition>> newPartit
+	/**
+	 * Set the {{@link #bytesPerSecondMax}} per consumer using the global bytes per second.
+	 */
+	public void setBytesPerSecondMax() {
```
Is this intended to be public or is that a mistake?  It looks like this should be private.

glaksh100 replied:
Not intended. Fixed.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -482,6 +511,67 @@ void reassignPartitions(List<KafkaTopicPartitionState<TopicPartition>> newPartit
+	/** Create RateLimiter with a rate of {{@link #bytesPerSecondMax}}.
+	 *
+	 */
+	private void createRateLimiter() {
```
Suggestion: Call this getRateLimiter() or getOrCreateRateLimiter() just to help with readability.

glaksh100 replied:
Updated to getOrCreateRateLimiter().

jgrier reviewed (approved):
I left a couple more comments @glaksh100 but this looks much better.  I don't have a strong opinion on trying to merge as is (after minor fixes I mentioned) or refactoring the KafkaConsumer for easier extension and then implementing this as a customization for Lyft.

Personally, I think this is a useful feature in general so maybe we should just try to merge it as is.  @tweise what do you think on this last point?

tweise commented:
@jgrier I had already expressed my preference on the dev@ thread and let's continue there.

@glaksh100 do we have a JIRA in Flink for this?

We should probably include this change into the upcoming build that needs to be made for the global accumulator, even if we have to back it out / augment it later due to non-concluded upstream discussion.

lrao100 committed:
Review updates and edge case fixes

glaksh100 commented:
@jgrier Addressed comments and also added a couple of conditionals on edge cases namely:
- Key in the Consumer Record being null (which is an allowed value)
- To ensure rateLimiter.acquire() is called with a value > 0

@tweise FLINK JIRA: https://issues.apache.org/jira/browse/FLINK-11501

Will create this PR against apache/flink

tweise commented:
@glaksh100 you are welcome to create a PR of course, but that isn't the recommended way to work in the Flink community and in ASF project generally. 

Please see https://flink.apache.org/contribute-code.html

Changes that are properly discussed upfront will be much more welcome than something popping out of nowhere. The mailing list discussion has not been concluded, some alternative ideas were brought up. I would suggest to respond to those and make a case for the approach you are proposing.

lrao100 committed:
Add a RateLimitingConfig and calculate local consumer ratelimit from a global limit

jgrier commented on flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer010.java:
```diff
@@ -69,6 +70,16 @@
+	private static final boolean DEFAULT_USE_RATELIMITING = false;
+
+	/** Default value for ratelimit which is default max.partition.fetch.bytes. **/
+	private static final long DEFAULT_MAX_BYTES_PER_SECOND = 1 * 1024 * 1024L;
```
This actually makes for a very low default rate limit.  Would it better to default to something much higher?

jgrier replied:
In fact I suggest no default.  If the feature is enabled and the rate limit isn't set it should throw an exception.

jgrier replied:
I also suggest you move this default to RateLimitingConfig.

glaksh100 replied:
No default is better, agreed.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/RateLimitingConfig.java:
```diff
@@ -0,0 +1,55 @@
+	 * @param consumerPrefix consumer name.
+	 * @return fully defined property with rate limiting flag.
+	 */
+	public static String getRatelimitFlag(String consumerPrefix) {
```
Make this return boolean and handle any defaulting, etc.  That makes this class more useful and eliminates code elsewhere.

glaksh100 replied:
Done.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/RateLimitingConfig.java:
```diff
@@ -0,0 +1,55 @@
+	 * @param consumerPrefix consumer name.
+	 * @return a fully defined property with max bytes per second.
+	 */
+	public static String getRatelimitMaxBytesPerSecond(String consumerPrefix) {
```
Make this return long and handle any defaulting, etc.  That makes this class more useful and eliminates code elsewhere.

glaksh100 replied:
Done.

jgrier commented on flink-connectors/flink-connector-kafka-0.10/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer010.java:
```diff
@@ -259,4 +272,25 @@ protected AbstractPartitionDiscoverer createPartitionDiscoverer(
+	 * @param runtimeContext
+	 * @param properties
+	 */
+	public void setRateLimitingProperties(
```
Move all of this logic to RateLimitingConfig.

glaksh100 replied:
Done.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer09.java:
```diff
@@ -311,4 +321,25 @@ private static void setDeserializer(Properties props) {
+	 * @param runtimeContext
+	 * @param properties
+	 */
+	public void setRateLimitingProperties(
```
Move all of this logic to RateLimitingConfig.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/internal/KafkaConsumerThread.java:
```diff
@@ -68,17 +69,14 @@
 
 	/** Default value for max bytes per second per consumer. **/
-	private static final long DEFAULT_BYTES_PER_SECOND_MAX = 1 * 1024 * 1024;
+	private static final long DEFAULT_BYTES_PER_SECOND_MAX = 1 * 1024 * 1024L;
```
Get rid of this.  Handle in RateLimitingConfig.

glaksh100 replied:
Done.

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaConsumer09.java:
```diff
@@ -82,6 +83,15 @@
+	/** Configuration to set consumer prefix for ratelimiting. **/
+	private static final String CONSUMER_PREFIX = "kafka";
+
+	/** Default value for ratelimit flag. **/
```
Move all of this to RateLimitingConfig.

glaksh100 replied:
Done.

jgrier reviewed (changes requested):
There's a lot of duplicate code here.   I suggest factoring all of that out into RateLimitingConfig.  Try to encapsulate ALL of the complexity of configuring rate limiting to one spot.  Config, defaults, translation from global to local limits, etc.  This will make the rest of the code much cleaner and remove all that duplication.

lrao100 committed:
Refactor RateLimitingConfig

lrao100 committed:
Update consumer 010

jgrier commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/RateLimitingConfig.java:
```diff
@@ -0,0 +1,115 @@
+ * A generic RateLimitingConfig that can be used to set ratelimiting properties across
+ * consumers.
+ */
+public class RateLimitingConfig {
```
De-static-ify this class, please.  @glaksh100 and I talked offline about this.  Just noting here for visibility.  This class will basically become a Factory that will build a already configured RateLimiter.  Easy peasy.

tweise replied:
If you can configure that as consumer property then this would also remove the Guava dependency and allow others to plug in their own implementation.

lrao100 committed:
Convert config class to factory

lrao100 committed:
Remove defaults from KafkaConsumerThread

glaksh100 commented:
I converted the Config class to a Factory that also creates the RateLimiter. The changes are more now since the signature of the `Kafka09Fetcher`, `Kafka010Fetcher`, `KafkaConsumerThread` constructors, have changed to include an additional parameter for the `RateLimiterFactory`.

@tweise I have not yet addressed the suggestion to allow for creating a non-Guava RateLimiter, which I think will be pretty useful. I'll look into that next.

jgrier reviewed (commented):
👍

tweise commented on flink-connectors/flink-connector-kafka-0.9/src/main/java/org/apache/flink/streaming/connectors/kafka/config/RateLimiterFactory.java:
```diff
@@ -0,0 +1,101 @@
+public class RateLimiterFactory {
+
+	/** Flag that indicates if ratelimiting is enabled. */
+	private static final String RATELIMITING_FLAG = "consumer.ratelimiting.enabled";
```
I don't think there is a need for any of these property keys. The user, if rate limiting is required, can set a rate limiter object that is already configured and encapsulates how rate limiting is done. Rate limiter should be just an interface and users can define their own implementation.

tweise commented:
See comment above regarding rate limiter interface. This is the wrong place to have these discussions and iterations, this belongs on the Flink mailing list thread.

Will merge this for internal use.
